### PR TITLE
ann chantal scrabble 

### PR DESCRIPTION
**Pull request by gofarann**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|   We shared equal time in both roles. We also likely shared equal amounts of time writing spec files and code files. The pomodoro timer (30min/30min) helped us track role changes. 	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  It's possible that you were trying to allow us to practice our skills accessing methods across classes since in the "real world" we might have to do similar things and work with multiple teams. Fundamentally, it is important to call class methods on Scoring because we need to know what is happening to the score as a class vs. for each and every instance of score. Instances of score would return no meaningful information.    |
|   Describe an example of a test you wrote for a _nominal case_. |   In the spec file for Tilebag, we wrote a test to see if the correct number of tiles remained in the bag after n.times of draw tiles was executed.	|
|   Describe an example of a test you wrote for an _edge case_. |   In the spec file for Tilebag, we wrote a test to see if an argument error would be raised if the requested number of tiles for draw tiles is larger than the number remaining in the bag. 	|
|   What was your final test coverage percentage?	|   100%	|
|   Is there a specific piece of code you'd like feedback on?	|   The draw_tiles method on the Scrabble Tilebag class	|

### CODE REVIEW

File: lib/scrabble_tilebag.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      ]
+      @player_bag = []
+    end
+
+    def draw_tiles(num)
+
+      raise ArgumentError.new "Not enough tiles left in the pile." if num > @bag.length
+      player_bag = []
+      num.times do
+        tile = @bag.sample # we can't use pop because it doesn't give us a random set
```
Comment (CheezItMan) [inline lib/scrabble_tilebag.rb:20]: Could you instead of used shuffle in `initialize` to have the array in a random order and then used `pop(num)` to remove the right number of items from the Array?

File: lib/scrabble_tilebag.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+         @player_bag << tile
+        player_bag << tile # player bag is used in the Player class where we don't need to track the same player bag, it's just a new one over and over
+        @bag.slice!(@bag.index(tile)) # we use slice vs delete because it will only delete one instance of the letter rather than all instances of that letter
```
Comment (CheezItMan) [inline lib/scrabble_tilebag.rb:23]: Good that you noted that delete will remove all occurrences, and slice will work, but you could also used,

`@bag.shuffle!`
Then 
`return @bag.slice!(num)`

File: lib/scrabble.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      %w[q z]=> 10
+    }
+
+    def self.score word
+
+      letter_values = []
+
+      raise ArgumentError.new "Error. Please enter a string." if word.class != String
+
+      raise ArgumentError.new "Error. Please enter a string of letters." if word.match(/^[[:alpha:]]+$/) == nil
```
Comment (CheezItMan) [inline lib/scrabble.rb:25]: Good use of Regex!

File: lib/scrabble.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+      raise ArgumentError.new "Error. Please enter a string that is less than 7 characters." if word.length > 7
```
Comment (CheezItMan) [inline lib/scrabble.rb:27]: Should be `Error. Please enter a string that is less than or equal to 7 characters.`

File: lib/scrabble.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    # constant that is a hash of the point values of the letters
+    SCOREBOARD = {
+      %w[a e i o u l n r s t]=>1,
+      %w[d g]=> 2, %w[b c m p]=>3,
+      %w[f h v w y]=> 4,
+      %w[k]=> 5,
+      %w[j x]=> 8,
+      %w[q z]=> 10
+    }
```
Comment (CheezItMan) [inline lib/scrabble.rb:17]: Good to use a hash, but it might have been easier in your `score` method if you had done:
```
SCOREBOARD = {
  'a' => 1,
  'e' => 1,
  'f' => 4,
  ...
}
```

File: lib/scrabble.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      raise ArgumentError.new "Error. Please enter a string that is less than 7 characters." if word.length > 7
+
+      word_array = word.downcase.scan /\w/ # regex that checks to make sure the string consists of only letters
+
+      # this checks the value of our letters
+      SCOREBOARD.each do |key, value|
+        word_array.each do |letter|
+          key.include?(letter) ? letter_values << value : letter_values << 0
+        end
+      end
```
Comment (CheezItMan) [inline lib/scrabble.rb:36]: This seems a little convoluted with nested `each` loops.  You could either simplify it using the hash I outlined above and `map` or use a `select` in the loop you have above to select all the letters inside `word_array` that appear in key and multiply the number by the value.

File: lib/scrabble.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      points_count = letter_values.inject { |points, n| points + n }
+
+      word_array.size < 7 ? points_count : points_count + 50 # adds bonus points for 7 letter words
+    end
+
+    def self.highest_score_from array
+
+      score = []
+
+      score << (array.map { |word| Scoring.score(word) }).first # score for the entire word
```
Comment (CheezItMan) [inline lib/scrabble.rb:47]: Score is always going to be just 1 element here, why an Array?

File: lib/scrabble.rb [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      winning_words = []
+      losing_words = []
+
+      array.each do |word|
+        score.max == Scoring.score(word) ? winning_words << word : losing_words << word
+      end
+
+      if winning_words.length > 1 ##
+        winning_words.any? { |word| word.length == 7 } ? winning_words.find { |word| word.length == 7 } : winning_words.min_by { |word| word.length}
+      end
```
Comment (CheezItMan) [inline lib/scrabble.rb:58]: I see what you're doing here, but you're doing a lot of work very compactly.  It might be easier to follow if you didn't use the ternary operator here.

File: lib/scrabble_player.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    def won?
+      @score_tracker.inject { |sum, n| sum + n } > 100 ? true : false #determines winner based on number of points
+    end
+
+
+    def play(word)
+
+      raise ArgumentError.new "Error. You cannot play an empty word. Please try again." if word =~ /^$|\s+/
+
+      raise ArgumentError.new "Error. You cannot play anymore. Please restart game." if @score_tracker.inject { |sum, n| sum + n } > 100
```
Comment (CheezItMan) [inline lib/scrabble_player.rb:31]: Couldn't you simply call `won?` here?

Also it **Should** be returning false instead of raising an `ArgumentError`.  Check the instructions

**Review by CheezItMan — COMMENTED** [review]
I've left some specific feedback, especially on `draw_tiles`

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   It's great that you were able to collaborate using the pomodoro timer to keep on track.  For **Score** we had you make it a class method because we don't need, "Score" objects running around.  We simply need to calculate a score for an individual word and return the answer.  No data needs to be remembered.  I've left feedback in your PR on the `draw_tiles` method.	|
|   Both Teammates contributed to the codebase	|   Check, good alternation of roles	|
|   Regular Commits with meaningful commit messages 	|   Check	|
|  **Scoring class**	|   Check, Good use of Regex to verify input.	|
|  `score` method	|   Check, but it's very convoluted.  See my comments in the PR code.  I think you got a little too far down the rabbit hole on a few, and a more easy-to-use hash of scores for letters would have simplified things more. 	|
|   Uses appropriate data structure to store the letter score	|   Not bad, but it makes the code you write in score, very dense and harder to follow.  See my comments in the PR Code.  	|
|   Appropriately handles edge cases	|   Good, but you missed checking the 50-point bonus for 7-letter-word-score	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   Missing	|
|  `highest_scoring_word` method	|   Check	|
|   Appropriately handles edge cases	|   Check	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   Missing a test for empty Array	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   Check	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   Check	|
|  Returns `true` or `false` from `won?` method | Check  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |    Check, except no test for empty strings |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   Check, nice use of an Array, just remember `shuffle`!	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   Check, but very complex, see my notes in the code.	|
| `tiles_remaining` method returns a count	|   Check	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   What happens if there are no more tiles to draw?	|

## Summary

Not bad!  You went a little deep in the weeds with very complex code, but it seems like you were trying to get practice using Enumerable methods.  I've left some comments in the code of your PR.  I like how you were able to split out the coding responsibilities using the timer.
